inputremapper/injection/consumers/mapping_handler.py

```diff
@@ -0,0 +1,379 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+# input-remapper - GUI for device specific keyboard mappings
+# Copyright (C) 2022 sezanzeb <proxima@sezanzeb.de>
+#
+# This file is part of input-remapper.
+#
+# input-remapper is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# input-remapper is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with input-remapper.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Consumer base class.
+
+Can be notified of new events so that inheriting classes can map them and
+inject new events based on them.
+"""
+import asyncio
+import evdev
+
+from typing import Dict, Tuple, List, Protocol
+from evdev.ecodes import EV_ABS
+
+from inputremapper import utils
+from inputremapper import exceptions
+from inputremapper.logger import logger
+from inputremapper.key import Key
+from inputremapper.mapping import Mapping
+from inputremapper.system_mapping import system_mapping
+from inputremapper.injection.macros.macro import Macro
+from inputremapper.injection.macros.parse import parse, is_this_a_macro
+from inputremapper.injection.global_uinputs import global_uinputs
+
+
+def copy_event(event: evdev.InputEvent) -> evdev.InputEvent:
+    return evdev.InputEvent(
+        sec=event.sec,
+        usec=event.usec,
+        type=event.type,
+        code=event.code,
+        value=event.value,
+    )
+
+
+class ContextProtocol(Protocol):
+    """the parst from context needed for macros"""
+
+    mapping: Mapping
+    last_btn_down_event: Tuple[int, int]
+    last_btn_up_event: Tuple[int, int]
+
+
+class CombinationSubHandler(Protocol):
+    """Protocol any handler which can be triggered by a combination must implement"""
+
+    @property
+    def active(self) -> bool:
+        ...
+
+    async def notify(self, event: evdev.InputEvent) -> bool:
+        ...
+
+    async def run(self) -> None:
+        ...
+
+
+class MappingHandler(Protocol):
+    """the protocol a mapping handler must follow"""
+
+    def __init__(self, config: Dict[str, int], context: ContextProtocol):
+        ...
+
+    async def notify(
+        self,
+        event: evdev.InputEvent,
+        source: evdev.InputDevice = None,
+        forward: evdev.UInput = None,
+        supress: bool = False,
+    ) -> bool:
+        ...
+
+    async def run(self) -> None:
+        ...
+
+
+class CombinationHandler:
+    """keeps track of a combination and notifies a sub handler
+
+    adheres to the MappingHandler protocol
```

Comment:
why is this not inheriting from `MappingHandler`?

Reply:
That was my first approach, but it turns out that the MappingHandler as a super class is not very useful. I ended up overriding pretty much all of its methods with every handler. It didn't implement any useful logic. 

I made the MappingHandler a Protocol class: https://www.python.org/dev/peps/pep-0544/
It serves as a useful type-hint, in the sens that any object which has the in `MappingHandler` specified properties and methods is considered to be a `MappingHandler` by most IDEs and type-checking.

I should probably rename it to something like `TopLevelHandler`, to make clear that this handler might be directly exposed to the context, and therefor must implement the methods and properties as described by the protocol.

A handler like the `KeyHandler` or `MacroHandler` does not follow the `MappingHandler` protocol if we try to add such a handler to the `context._handlers` dict we will get a type warning by the IDE